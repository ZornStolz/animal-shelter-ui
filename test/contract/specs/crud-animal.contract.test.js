import { provider } from "../config/init-pact";
import { Matchers } from "@pact-foundation/pact";
import { AnimalController } from "../../../controllers";

describe("Animal Service", () => {
  describe("When a request to list all animals is made", () => {
    beforeAll(async () => {
      await provider.setup();
      await provider.addInteraction({
        uponReceiving: "a request to list all animals",
        state: "has animals",
        withRequest: {
          method: "GET",
          path: "/animals",
        },
        willRespondWith: {
          status: 200,
          body: Matchers.eachLike({
            name: Matchers.like("manchas"),
            breed: Matchers.like("Bengali"),
            gender: Matchers.like("Female"),
            vaccinated: Matchers.boolean(true),
          }),
        },
      });
    });

    test("should return a list with animals", async () => {
      const response = await AnimalController.list();

      expect(response.data).toMatchSnapshot();
      await provider.verify();
    });

    afterAll(() => provider.finalize());
  });

  describe("When a request to get an especific animal is made", () => {
    beforeAll(async () => {
      await provider.setup();
      await provider.addInteraction({
        uponReceiving: "a request to find an animal",
        state: "get animal by name",
        withRequest: {
          method: "GET",
          path: "/animals/manchas",
        },
        willRespondWith: {
          status: 200,
          body: Matchers.eachLike({
            name: Matchers.like("manchas"),
            breed: Matchers.like("Bengali"),
            gender: Matchers.like("Female"),
            vaccinated: Matchers.boolean(true),
            vaccines: ["Leptospirosis", "Parvovirus"],
          }),
        },
      });
    });

    test("The Sought animal was found and has the correct data", async () => {
      const response = await AnimalController.getAnimal("manchas");

      expect(response.data).toMatchSnapshot();
      await provider.verify();
    });

    afterAll(() => provider.finalize());
  });

  describe("When a request to post an animal is made", () => {
    beforeAll(async () => {
      await provider.setup();
      await provider.addInteraction({
        uponReceiving: "a request to post an animal",
        state: "add animal",
        withRequest: {
          method: "POST",
          path: "/animals",
        },
        willRespondWith: {
          status: 201,
          body: Matchers.eachLike({
            id: Matchers.like(69),
            name: Matchers.like("manchas"),
            breed: Matchers.like("Bengali"),
            gender: Matchers.like("Female"),
            vaccinated: Matchers.boolean(true),
            vaccines: ["Leptospirosis", "Parvovirus"],
          }),
        },
      });
    });

    test("The animal was created and has the correct data", async () => {
      const animal = {
        name: "manchas",
        breed: "Bengali",
        gender: "Female",
        vaccinated: true,
        vaccines: ["Leptospirosis", "Parvovirus"],
      };
      const response = await AnimalController.register(animal);

      expect(response.data).toMatchSnapshot();
      await provider.verify();
    });

    afterAll(() => provider.finalize());
  });

  describe("When a request to update an animal is made", () => {
    beforeAll(async () => {
      await provider.setup();
      await provider.addInteraction({
        uponReceiving: "a request to update an animal",
        state: "update animal",
        withRequest: {
          method: "PUT",
          path: "/animals/manchas",
        },
        willRespondWith: {
          status: 200,
          body: Matchers.eachLike({
            id: Matchers.like(69),
            name: Matchers.like("manchas"),
            breed: Matchers.like("Bengali"),
            gender: Matchers.like("Female"),
            vaccinated: Matchers.boolean(true),
            vaccines: ["Leptospirosis", "Parvovirus"],
          }),
        },
      });
    });

    test("The animal was updated and has the correct data", async () => {
      const response = await AnimalController.updateAnimal("manchas");

      expect(response.data).toMatchSnapshot();
      await provider.verify();
    });

    afterAll(() => provider.finalize());
  });

  describe("When a request to delete an animal is made", () => {
    beforeAll(async () => {
      await provider.setup();
      await provider.addInteraction({
        uponReceiving: "a request to delete an animal",
        state: "delete animal",
        withRequest: {
          method: "DELETE",
          path: "/animals/manchas",
        },
        willRespondWith: {
          status: 204,
        },
      });
    });

    test("The animal was deleted and the status is 204", async () => {
      const response = await AnimalController.delete("manchas");
      expect(response).toMatchSnapshot();
      await provider.verify();
    });

    afterAll(() => provider.finalize());
  });
});
