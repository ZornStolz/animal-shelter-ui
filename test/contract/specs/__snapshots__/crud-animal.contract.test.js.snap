// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Animal Service When a request to delete an animal is made The animal was deleted and the status is 204 1`] = `
Object {
  "config": Object {
    "adapter": [Function],
    "baseURL": "http://localhost:8080",
    "data": undefined,
    "headers": Object {
      "Accept": "application/json, text/plain, */*",
    },
    "maxBodyLength": -1,
    "maxContentLength": -1,
    "method": "delete",
    "timeout": 0,
    "transformRequest": Array [
      [Function],
    ],
    "transformResponse": Array [
      [Function],
    ],
    "transitional": Object {
      "clarifyTimeoutError": false,
      "forcedJSONParsing": true,
      "silentJSONParsing": true,
    },
    "url": "animals/Manchas",
    "validateStatus": [Function],
    "xsrfCookieName": "XSRF-TOKEN",
    "xsrfHeaderName": "X-XSRF-TOKEN",
  },
  "data": "",
  "headers": Object {},
  "request": XMLHttpRequest {},
  "status": 204,
  "statusText": "No Content ",
}
`;

exports[`Animal Service When a request to get an especific animal is made The Sought animal was found and has the correct data 1`] = `
Object {
  "breed": "Bengali",
  "gender": "Female",
  "name": "Manchas",
  "vaccinated": true,
  "vaccines": Array [
    "Leptospirosis",
    "Parvovirus",
  ],
}
`;

exports[`Animal Service When a request to list all animals is made should return a list with animals 1`] = `
Array [
  Object {
    "breed": "Bengali",
    "gender": "Female",
    "name": "Manchas",
    "vaccinated": true,
  },
]
`;

exports[`Animal Service When a request to post an animal is made The animal was created and has the correct data 1`] = `
Object {
  "breed": "Bengali",
  "gender": "Female",
  "name": "Manchas",
  "vaccinated": true,
  "vaccines": Array [
    "Leptospirosis",
    "Parvovirus",
  ],
}
`;

exports[`Animal Service When a request to update an animal is made The animal was updated and has the correct data 1`] = `
Object {
  "breed": "Bengali",
  "gender": "Female",
  "name": "Manchas",
  "vaccinated": true,
  "vaccines": Array [
    "Leptospirosis",
    "Parvovirus",
  ],
}
`;
